qlight/server.go

```diff
@@ -0,0 +1,186 @@
+	txHash := common.BytesToEncryptedPayloadHash(encryptedPayloadHash)
+	_, _, privateTx, extra, err := p.ptm.Receive(txHash)
+	if err != nil {
+		return nil, err
+	}
+	// we're not party to this transaction
+	if privateTx == nil {
+		return nil, nil
+	}
+	if p.privateStateManager.NotIncludeAny(psm, extra.ManagedParties...) {
+		return nil, nil
+	}
+
+	extra.ManagedParties = psm.FilterAddresses(extra.ManagedParties...)
+
+	ptd := PrivateTransactionData{
+		Hash:     &txHash,
+		Payload:  privateTx,
+		Extra:    extra,
+		IsSender: false,
+	}
+	if len(psm.Addresses) == 0 {
+		// this is not an MPS node so we have to ask tessera
+		ptd.IsSender, _ = p.ptm.IsSender(txHash)
+	} else {
+		// this is an MPS node so we can speed up the IsSender logic by checking the addresses in the private state metadata
+		ptd.IsSender = !psm.NotIncludeAny(extra.Sender)
+	}
+
+	return &ptd, nil
+}
+
+type authProviderImpl struct {
+	privateStateManager mps.PrivateStateManager
+	authManagerProvider AuthManagerProvider
+	authManager         security.AuthenticationManager
+	enabled             bool
+}
+
+func NewAuthProvider(privateStateManager mps.PrivateStateManager, authManagerProvider AuthManagerProvider) AuthProvider {
+	return &authProviderImpl{
+		privateStateManager: privateStateManager,
+		authManagerProvider: authManagerProvider,
+		enabled:             false,
+	}
+}
+
+func (a *authProviderImpl) Initialize() error {
+	if a.authManagerProvider != nil {
+		a.authManager = a.authManagerProvider()
+		if a.authManager == nil {
+			return nil
+		}
+		authEnabled, err := a.authManager.IsEnabled(context.Background())
+		if err != nil {
+			return err
+		}
+		a.enabled = authEnabled
+	}
+	return nil
+}
+
+func (a *authProviderImpl) Authorize(token string, psi string) error {
+	if !a.enabled {
+		return nil
+	}
+
+	authToken, err := a.authManager.Authenticate(context.Background(), token)
+	if err != nil {
+		return err
+	}
+	PSI := types.PrivateStateIdentifier(psi)
+	// check that we have access to the relevant PSI
+	psiAuth, err := multitenancy.IsPSIAuthorized(authToken, PSI)
+	if err != nil {
+		return err
+	}
+	if !psiAuth {
+		return fmt.Errorf("PSI not authorized")
+	}
+	// check that we have access to  qlight://p2p , rpc://eth_*
+	// TODO figure out what other entitlements we want to check here
+	qlightP2P := false
+	rpcETH := false
+	for _, ga := range authToken.GetAuthorities() {
+		if ga.GetRaw() == "p2p://qlight" {
+			qlightP2P = true
+		}
+		if ga.GetRaw() == "rpc://eth_*" {
+			rpcETH = true
+		}
+	}
+	if !qlightP2P || !rpcETH {
+		log.Error("The P2P token does not have the necessary authorization", "p2p", qlightP2P, "rpcETH", rpcETH)
+		return fmt.Errorf("The P2P token does not have the necessary authorization p2p=%v rpcETH=%v", qlightP2P, rpcETH)
+	}
+	// try to resolve the PSI
+	_, err = a.privateStateManager.ResolveForUserContext(rpc.WithPrivateStateIdentifier(context.Background(), PSI))
+	if err != nil {
+		return err
```

Comment:
wrap error:
```suggestion
		return fmt.Errorf("resolve PSI: %w", err)
```
or simply `return err` if no wrapping

Reply:
handled